### ISSUE TITLE
Is there a way to print a schema with references resolved?

### ISSUE THREAD

**Issue by kelvinpho** [java-json-tools/json-schema-validator#41]
Hi,

I'm not having any issues with SchemaNode and validation, but I would like a way to print my SchemaNode with $ref resolved.  An example of this is on http://www.jsonschema.net/ where they can pretty print the Json Schema in JSON format.

Is there any way to do this with json-schema-validator's SchemaNode or SchemaTree?

Thanks!

**Comment by olgabo** [comment]
I can't see RefExpander anymore in the master tree? Has it been removed

**Comment by queenaqian** [comment]
I have the same requirement, is it a release feature to print a complete schema?

**Comment by simmosn** [comment]
> I have the same requirement, is it a release feature to print a complete schema?

Have the same question also, did you get an answer on this?

**Comment by huggsboson** [comment]
I think where they landed, after reading the discussion is that you need to implement your own as a Schema walker.  I didn't see them merge anything into the tree to do it.  Happy to take a default implementation of it if any of you do it.

**Comment by neerajsu** [comment]
@olgabo , @queenaqian , @simmosn , @huggsboson 

@fge actually provided ResolvingSchemaWalker which does everything what the OP needs out of the box. It took me to while to understand by looking at the source code, but it is very straightforward. Here's the code snippet for you to use.

```
URL url = classLoader.getResource("/some/schema/file/path/in/resource");
        JsonNode jsonNode = JsonLoader.fromURL(url);
        SchemaTree schemaTree = new CanonicalSchemaTree(jsonNode);
        ResolvingSchemaWalker resolvingSchemaWalker = new ResolvingSchemaWalker(schemaTree);
        SchemaExpander schemaTreeSchemaListener = new SchemaExpander(schemaTree);
        resolvingSchemaWalker.walk(schemaTreeSchemaListener, new ConsoleProcessingReport());
        SchemaTree resolvedSchemaTree = schemaTreeSchemaListener.getValue();
```

resolvedSchemaTree should now have all "$ref" in your schema resolved. Hope that answers your question.

**Comment by danham-slalom** [comment]
@neerajsu is the ResolvingSchemaWalker still available? I'm not able to find it in the repository. Any idea if it has been removed with more recent version changes? [The latest version I can find mention of the ResolvingSchemaWalker is 1.1.8](https://jar-download.com/maven-repository-class-search.php?search_box=ResolvingSchemaWalker) but the [latest version still available on maven central is 1.5](https://mvnrepository.com/artifact/com.github.fge/json-schema-validator).

**Comment by neerajsu** [comment]
> @neerajsu is the ResolvingSchemaWalker still available? I'm not able to find it in the repository. Any idea if it has been removed with more recent version changes? [The latest version I can find mention of the ResolvingSchemaWalker is 1.1.8](https://jar-download.com/maven-repository-class-search.php?search_box=ResolvingSchemaWalker) but the [latest version still available on maven central is 1.5](https://mvnrepository.com/artifact/com.github.fge/json-schema-validator).

My comment is almost 2 years old. At that time, ResolvingSchemaWalker was experimental. So it's probably renamed and/or incorporated into main code. You'll have to dig into the latest source code to figure out. Im sure the functionality exists in the latest code.
